### DISCUSSION TITLE
Error serializing response with both `response_class` and `response_model`

### DISCUSSION THREAD

**Discussion by kunansy** [fastapi/fastapi#9739] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import APIRouter
from fastapi.responses import HTMLResponse, JSONResponse
from pydantic import BaseModel

router = APIRouter(
    prefix="/system",
    tags=['system'],
    default_response_class=HTMLResponse
)

class TestResponse(BaseModel):
    test: int

@router.post('/test',
             # I guess response model should replace response class
             response_model=TestResponse)
async def test():
    return {
        "test": 123
    }
```

### Description

Hi folks 👋.

I encountered a problem, calling the example code gives an error:
```
File "/.../lib/python3.11/site-packages/starlette/responses.py", line 55, in __init__
    self.body = self.render(content)
                ^^^^^^^^^^^^^^^^^^^^
  File "/.../lib/python3.11/site-packages/starlette/responses.py", line 64, in render
    return content.encode(self.charset)
           ^^^^^^^^^^^^^^
AttributeError: 'dict' object has no attribute 'encode'
```

It would be OK If I set `response_class=JSONResponse`. 
```
@router.post('/test',
             response_model=TestResponse,
             response_class=JSONResponse)
async def test(): ...
```

So, I guess when I set `response_model` the default `response_class` has to be replaced. I have to set both `response_class` and `response_model` to make it work, it looks weird and makes me sad 😞.

Where am I wrong? 🧐

### Operating System

macOS

### FastAPI Version

0.97.0

### Python Version

Python 3.11.4

**Comment by jgould22** [comment, marked as answer]
You will have to set the response class because you set the default on your router to HTMLResponse.

When you do that the router is going to try and encode that dict as utf-8. See the code here https://github.com/encode/starlette/blob/master/starlette/responses.py#L58

You are explicitly telling the router not to use its default (seen here https://github.com/tiangolo/fastapi/blob/master/fastapi/routing.py#L493 ) so you must re-set it back to json as you are doing.
